detection_rules/rule_loader.py

```diff
@@ -4,45 +4,209 @@
+        # https://github.com/uiri/toml/issues/152
+        # might also be worth looking at https://github.com/sdispater/tomlkit
+        with io.open(str(path.resolve()), "r", encoding="utf-8") as f:
+            toml_dict = pytoml.load(f)
+            self._toml_cache[path] = toml_dict
+            return toml_dict
+
+    def _get_paths(self, directory: Path, recursive=True) -> List[Path]:
+        if recursive:
+            paths = glob.glob(str(directory / "**" / "*.toml"), recursive=True)
+        else:
+            paths = glob.glob(str(directory / "*.toml"))
+
+        return [Path(path) for path in paths]
+
+    def load_file(self, path: Path) -> TOMLRule:
         try:
-            return f(*args, **kwargs)
-        finally:
-            load_rules.clear()
+            path = path.resolve()
+
+            # use the default rule loader as a cache.
+            # if it already loaded the rule, then we can just use it from that
+            if self.__default is not None and self is not self.__default and path in self.__default.file_map:
+                rule = self.__default.file_map[path]
+                self.add_rule(rule)
+                return rule
+
+            obj = self._deserialize_toml(path)
+            return self.load_dict(obj, path=path)
+        except Exception:
+            print(f"Error loading rule in {path}")
+            raise
+
+    # noinspection PyShadowingBuiltins
+    def load_directory(self, directory: Path, recursive=True, toml_filter: Optional[Callable[[dict], bool]] = None):
+        paths = self._get_paths(directory, recursive=recursive)
+        if toml_filter is not None:
+            paths = [path for path in paths if toml_filter(self._deserialize_toml(path))]
 
-    return wrapped
+        self.load_files(paths)
 
+    def load_by_id(self, rule_id: definitions.UUIDString,
+                   directory: Path = Path(RULES_DIR), recursive=True) -> Optional[TOMLRule]:
+        """Search for a rule by ID and load it."""
+        if rule_id in self.id_map:
+            return self.id_map[rule_id]
+        elif self.__default is not None and rule_id in self.id_map:
+            return self.__default.id_map[rule_id]
 
-def reset():
-    """Clear all rule caches."""
-    load_rule_files.clear()
-    load_rules.clear()
-    get_rule.clear()
-    filter_rules.clear()
+        # otherwise, go looking
+        self.load_directory(directory, toml_filter=dict_filter(rule__rule_id=rule_id), recursive=recursive)
+
+        # now check if it was found
+        return self.id_map.get(rule_id)
+
+    def freeze(self):
+        """Freeze the rule collection and make it immutable going forward."""
+        self.frozen = True
+
+    @classmethod
+    def default(cls):
+        """Return the default rule collection, which retrieves from rules/."""
+        if cls.__default is None:
+            collection = RuleCollection()
+            collection.load_directory(Path(RULES_DIR))
```

Comment:
this is an awesome idea! Should this be:

```suggestion
            collection.load_directory(Path(RULES_DIR))
            collection.freeze = True
```

or add a `freeze` method and

```suggestion
            collection.load_directory(Path(RULES_DIR))
            collection.freeze()
```

Reply:
oh i forgot to call the `freeze()` method, good catch.
the rationale here, is that if you load the default collection, you can't change it. but you can always create your own. this makes the caching code effectively disappear. there's only ever one default